proposals/attach-to-property.md

```diff
@@ -0,0 +1,142 @@
+# Attach to delegate property object
+
+* **Type**: Design proposal
+* **Author**: Ilya Ryzhenkov
+* **Contributors**:
+* **Status**:
+* **Prototype**: Was implemented and then disabled
+
+## Summary
+
+Recognize special operator function available for delegate object and call it right after it has been stored
+into a field for delegation.
+
+## Glossary
+
+- "delegate instance": instance on which property's get/set accessors execute respective getValue/setValue calls.
+- "host instance": instance containing delegated properties as members, or being extended with extension property.
+- "attachTo": placeholder for the name of the operator function
+- "constructor" means either primary constructor, or each of secondary constructor if there is no primary constructor
+
+## Motivation / use cases
+
+Make delegate instance aware of host instance and property it was used with.
+
+- Register host instance in external table of managed entities, e.g. database schema, component container.
+- Register delegate properties in host instance with callback for "metadata", e.g. for database columns.
+- Call back into host instance with additional information, e.g. property metadata, delegate instance.
+  Can be used to communicate delegate instance back to the host instance for later use, by declaring member
+  extension `attachTo` function with receiver of delegate type.
+- Execute custom code before first access to property, e.g. `async` property that is being calculated
+  in the background, and blocks as needed when trying to use the value. With coroutines, we can suspend instead
+  of blocking.
+
+## Description
+
+Client code:
+```
+class HostType {
+   val name by DelegateType()
+}
+```
+
+Attach operator:
+```
+operator fun DelegateType.attachTo(instance: HostType, property: KProperty<*>) : Unit`
+```
+
+- No extra instances are created, all information is already available
```

Comment:
I was thinking about it, since we have "unused" return type of the function, but thinking about various consequences I discarded this idea. I will add some text about why this is not an option.